### ISSUE TITLE
object_can_type breaks on lazy attribute ISA checks under debugger

### ISSUE THREAD

**Issue by wchristian** [houseabsolute/Specio#6]
Found this by trying to fix perl-code-tidyall.

On instrospection it seems to use `$_[0]` fairly liberally in generated code for ISA checks on lazy attributes, instead of `$value`.

```
D:\cpan\perl-code-tidyall>perl -Ilib t\Basic.t
ok 1 - one file no plugins - error_count == 0
ok 2 - one file no plugins - foo.txt content
[...]

D:\cpan\perl-code-tidyall>perl -d -Ilib t\Basic.t

Loading DB routines from perl5db.pl version 1.39_11
Editor support available.

Enter h or 'h h' for help, or 'perldoc perldebug' for more help.

Attribute::Handlers::CODE(0x41f530c)(C:/strawberry/perl/lib/Attribute/Handlers.pm:251):
251:                    $global_phase++;
  DB<1> c
not ok 1 - test_basic died (isa check for "cache" failed: Can't call method "can" on unblessed reference at (eval 187)[C:/strawberry/perl/site/lib/Sub/Quote.pm:3] line 46.)
#   Failed test 'test_basic died (isa check for "cache" failed: Can't call method "can" on unblessed reference at (eval 187)[C:/strawberry/perl/site/lib/Sub/Quote.pm:3] line 46.)'
#   at t\Basic.t line 5.
#   (in Test::Code::TidyAll::Basic->test_basic)
not ok 2 - test_caching_and_backups died (isa check for "cache" failed: Can't call method "can" on unblessed reference at (eval 187)[C:/strawberry/perl/site/lib/Sub/Quote.pm:3] line 46.)
#   Failed test 'test_caching_and_backups died (isa check for "cache" failed: Can't call method "can" on unblessed reference at (eval 187)[C:/strawberry/perl/site/lib/Sub/Quote.pm:3] line 46.)'
#   at t\Basic.t line 5.
#   (in Test::Code::TidyAll::Basic->test_caching_and_backups)
```

**Comment by autarch** [comment]
Looks like the same issue as https://github.com/houseabsolute/DateTime.pm/issues/49

I'm not sure exactly what the problem is.

**Comment by wchristian** [comment]
The generated code has no reasonable expectation of working, unless one assumes XS magic being in play. You're trying to access `$value`, which has been copied into `@_`, inside a block by accessing `$_[0]`. However under normal conditions `@_` inside that block would've been set by List::Util, which has no reasonable way of accessing the outer `@_` without magic. In short, there's no way around but setting up proper, simple, pureperl saving of $_[0] into a named variable that you can then access in the List::Util block.

```
{
    my $isa_check_captures_for_cache = ${ $_[1]->{"\$isa_check_captures_for_cache"} };
    my $_UNQUOTED                    = ${ $_[1]->{"\$_UNQUOTED"} };
    my $_QUOTED                      = ${ $_[1]->{"\$_QUOTED"} };

    package Code::TidyAll;
    no warnings 'closure';

    sub cache {
        ( $_QUOTED, $_UNQUOTED ) if 0;

        # BEGIN quote_sub PRELUDE
        package Method::Generate::Accessor::_Generated;

        BEGIN {
            $^H = 1762;
            ${^WARNING_BITS} = "UUUUUUUUUUUUUUU\025";
            %^H = ();
        }

        # END quote_sub PRELUDE
        Carp::croak( "cache is a read-only accessor" ) if @_ > 1;
        exists $_[0]->{"cache"} ? $_[0]->{"cache"} :    #
          do {
            my $value = $_[0]->_build_cache;
            do {
                local $Method::Generate::Accessor::CurrentAttribute =
                  { init_arg => undef(), name => "cache", step => "isa check" };
                ( my $_error ), ( my $_old_error = $@ );
                (
                    eval {
                        ( $@ = $_old_error ),           #
                          (
                            do {
                                my $type              = ${ $isa_check_captures_for_cache->{"\$type"} };
                                my $message_generator = ${ $isa_check_captures_for_cache->{"\$message_generator"} };
                                local @_ = ( $value );

                                # BEGIN quote_sub PRELUDE

                                package Specio::Constraint::Role::Interface;

                                BEGIN {
                                    $^H = 2018;
                                    ${^WARNING_BITS} = "UUUUUUUUUUUUUUU\025";
                                    %^H = ();
                                }

                                # END quote_sub PRELUDE

                                ( Scalar::Util::blessed( $_[0] ) && List::Util::all { $_[0]->can( $_ ) } "get", "set" )
                                  or Specio::Exception->throw(
                                    message => $message_generator->( $_[0] ),
                                    type    => $type,
                                    value   => $_[0]
                                  );
                              }
                          ),
                          1;
                    }
                      or $_error = CORE::ref $@ ? $@ : "isa check for \"cache\" failed: " . $@
                );
                ( $@ = $_old_error ), ( defined $_error and CORE::die $_error );
            };
            $_[0]->{"cache"} = $value;
          }
    }
    $$_UNQUOTED = \&cache;
}
1;
```

**Comment by wchristian** [comment]
TBH, looking at the List::Util repo, nothing i see in there implies this should work. Do you remember what gave you the idea it should?

**Comment by autarch** [comment]
I have no idea why I thought this would work. The real problem is that it did work ;)

I've fixed this and will have a new release out tonight or tomorrow morning.

**Comment by wchristian** [comment]
Haha, excellent. I've also given LeoNerd a ping so he'll be aware and can maybe do something. :)

**Comment by autarch** [comment]
Fixed in Specio 0.32
